### PR TITLE
near: verify signatures through HostFunctionProvider

### PR DESCRIPTION
**Pull request by blasrodri**
Implement signature validations through `ed25519_recover` in `HostFunctionsProvider`

### CODE REVIEW

File: modules/src/clients/ics13_near/types.rs [hunk 1]
```diff
@@ -47,16 +47,14 @@ impl Signature {
...
         &self,
-        data: impl AsRef<[u8; 32]>,
-        public_key: PublicKey,
+        data: &[u8; 32],
+        public_key: &PublicKey,
     ) -> bool {
         match self {
-            Self::Ed25519(signature) => T::ed25519_recover(signature.as_ref(), data.as_ref())
-                .map(|key| key == public_key.0.as_ref())
+            Self::Ed25519(signature) => T::ed25519_recover(signature.as_ref(), data)
```
Comment (Wizdave97) [inline modules/src/clients/ics13_near/types.rs]: Not really sure about recovering ed25519 public keys from signatures.
Check this article for more context https://wiki.polkadot.network/docs/learn-keys.

That's why public key recovery for Schnorr signatures are not even implemented on substrate.

File: modules/src/clients/ics13_near/client_def.rs [hunk 2]
```diff
@@ -307,14 +308,15 @@ pub fn validate_light_block<H: HostFunctionsProvider>(
...
-        //     .verify::<H>(&H::sha256_digest(&approval_message), validator_public_key.clone())
-        // {
-        //     return Err(NearError::invalid_signature().into());
-        // }
+        let validator_public_key = &bp_stake_view.public_key;
+        let data = H::sha256_digest(&approval_message);
+        if !maybe_signature
+            .as_ref()
+            .unwrap()
+            .verify::<H>(&data, validator_public_key)
```
Comment (Wizdave97) [inline modules/src/clients/ics13_near/client_def.rs]: We could make this a host function?

File: modules/src/clients/ics13_near/types.rs (hunk 1 above)
Comment (Wizdave97) [reply modules/src/clients/ics13_near/types.rs]: Also mentioned in the last section here https://cryptobook.nakov.com/digital-signatures/eddsa-and-ed25519#:~:text=The%20produced%20digital%20signature%20is,%2B%2057%20bytes)%20for%20Ed448.

Comment (blasrodri) [reply modules/src/clients/ics13_near/types.rs]: I've done this in the PoC: https://github.com/ComposableFi/near-trustless-bridge/blob/b471d7fe7dc3bd692036b10b53ab465d12fa0750/near-lite-client/src/signature.rs#L9
Not sure why @seunlanlege chose to do ti through `ed25519_recover`.

File: modules/src/clients/ics13_near/client_def.rs (hunk 2 above)
Comment (blasrodri) [reply modules/src/clients/ics13_near/client_def.rs]: I have no problems with that. But let's figure out first https://github.com/ComposableFi/ibc-rs/pull/8#discussion_r883471726

File: modules/src/clients/ics13_near/types.rs (hunk 1 above)
Comment (blasrodri) [reply modules/src/clients/ics13_near/types.rs]: fixed in [2ee0220](https://github.com/ComposableFi/ibc-rs/pull/8/commits/2ee0220f5940fae191d17b8d198be4b67a7e6ee8)

File: modules/src/clients/ics13_near/client_def.rs (hunk 2 above)
Comment (blasrodri) [reply modules/src/clients/ics13_near/client_def.rs]: done in [93d3cb5](https://github.com/ComposableFi/ibc-rs/pull/8/commits/93d3cb5835b48ab3588e899e0e49bacb70a2e853)

**Review by seunlanlege — APPROVED** [review]

**Comment by seunlanlege** [comment]
looks good

**Pull request merged**